presto-main/src/main/java/com/facebook/presto/operator/JoinProbeBase.java

```diff
@@ -13,7 +13,25 @@
  */
 package com.facebook.presto.operator;
 
+import com.facebook.presto.spi.Page;
+import com.facebook.presto.spi.PageBuilder;
+import com.facebook.presto.spi.block.Block;
+import com.facebook.presto.spi.block.DictionaryBlock;
+
 public abstract class JoinProbeBase
     implements JoinProbe
 {
+    protected Page internalBuildDictionaryPage(int[] indices, int[] probeOutputChannels, PageBuilder sourcePageBuilder)
+    {
+        int sourceChannelCount = sourcePageBuilder.getChannelCount();
+        Block[] blocks = new Block[probeOutputChannels.length + sourceChannelCount];
+        int index = 0;
+        for (int i = 0; i < probeOutputChannels.length; i++) {
+            blocks[index++] = DictionaryBlock.mask(getPage().getBlock(probeOutputChannels[i]), indices);
+        }
+        for (int i = 0; i < sourceChannelCount; i++) {
+            blocks[index++] = sourcePageBuilder.getBlockBuilder(i).build();
```

Comment:
But you waste pretty little. The built `Page` is just a thin wrapper around array of blocks. You use blocks and waste this thin wrapper. On the pros side is cleaner API. With `PageBuilder` one can expect the method to actually _write to_ not _read from_ the `PageBuilder`.